lib/apiservers/engine/proxy/storage_proxy.go

```diff
@@ -0,0 +1,612 @@
+	switch numVolParams {
+	case 1:
+		VolumeID, err := uuid.NewUUID()
+		if err != nil {
+			return fields, err
+		}
+		fields.ID = VolumeID.String()
+		fields.Dest = volumeStrings[0]
+		fields.Flags = "rw"
+	case 2:
+		fields.ID = volumeStrings[0]
+		fields.Dest = volumeStrings[1]
+		fields.Flags = "rw"
+	case 3:
+		fields.ID = volumeStrings[0]
+		fields.Dest = volumeStrings[1]
+		fields.Flags = volumeStrings[2]
+	default:
+		// NOTE: the docker cli should cover this case. This is here for posterity.
+		return volumeFields{}, errors.InvalidBindError{Volume: volString}
+	}
+	return fields, nil
+}
+
+// processVolumeFields parses fields for volume mappings specified in a create/run -v.
+// It returns a map of unique mountable volumes. This means that it removes dupes favoring
+// specified volumes over anonymous volumes.
+func processVolumeFields(volumes []string) (map[string]volumeFields, error) {
+	volumeFields := make(map[string]volumeFields)
+
+	for _, v := range volumes {
+		fields, err := processVolumeParam(v)
+		log.Infof("Processed volume arguments: %#v", fields)
+		if err != nil {
+			return nil, err
+		}
+		volumeFields[fields.Dest] = fields
+	}
+	return volumeFields, nil
+}
+
+func finalizeVolumeList(specifiedVolumes, anonymousVolumes []string) ([]volumeFields, error) {
+	log.Infof("Specified Volumes : %#v", specifiedVolumes)
+	processedVolumes, err := processVolumeFields(specifiedVolumes)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("anonymous Volumes : %#v", anonymousVolumes)
+	processedAnonVolumes, err := processVolumeFields(anonymousVolumes)
+	if err != nil {
+		return nil, err
+	}
+
+	//combine all volumes, specified volumes are taken over anonymous volumes
+	for k, v := range processedVolumes {
+		processedAnonVolumes[k] = v
+	}
+
+	finalizedVolumes := make([]volumeFields, 0, len(processedAnonVolumes))
+	for _, v := range processedAnonVolumes {
+		finalizedVolumes = append(finalizedVolumes, v)
+	}
+	return finalizedVolumes, nil
+}
+
+func newVolumeCreateReq(name, driverName string, volumeData, labels map[string]string) (*models.VolumeRequest, error) {
+	if _, ok := SupportedVolDrivers[driverName]; !ok {
+		return nil, fmt.Errorf("error looking up volume plugin %s: plugin not found", driverName)
+	}
+
+	if !volumeNameRegex.Match([]byte(name)) && name != "" {
+		return nil, fmt.Errorf("volume name %q includes invalid characters, only \"[a-zA-Z0-9][a-zA-Z0-9_.-]\" are allowed", name)
+	}
+
+	req := &models.VolumeRequest{
+		Driver:     driverName,
+		DriverArgs: volumeData,
+		Name:       name,
+		Metadata:   make(map[string]string),
+	}
+
+	metadata, err := createVolumeMetadata(req, volumeData, labels)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Metadata[DockerMetadataModelKey] = metadata
+
+	if err := validateDriverArgs(volumeData, req); err != nil {
+		return nil, fmt.Errorf("bad driver value - %s", err)
+	}
+
+	return req, nil
+}
+
+func validateDriverArgs(args map[string]string, req *models.VolumeRequest) error {
```

Comment:
Seems to do more than validate but looks like it's a historically misnamed method so don't change it currently.